Fix pool listing on testnet

Anviking opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

#1794 


# Overview


- [x] Support pool-reregistration (instead of crashing)
- [x] Log pool retirements (But not doing anything else with them… yet)
- [x] Allow pools to have rewards but no stake (not sure why this happens 🤷‍♂️ ) Asked about it [here](https://input-output-rnd.slack.com/archives/CCRB7BU8Y/p1593085379334200)

# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

Anviking committed:
Allow pool re-registration

A new pool registration certificate may appear on-chain with the same
pool id (public key). This commit overwrites the existing certificate.

TODO: This won't work with rollbacks :thiking:

Anviking committed:
More detailed failure on metadata hash mismatches

Anviking committed:
wip: Log pool retirement certificates when they appear

TODO: Do something with them?

Anviking committed:
Allow pools to be in the rewards-reponse but have no stake

This seems to happen on the shelley_testnet. Not sure why.

KtorZ commented on lib/core/src/Cardano/Pool/DB/Sqlite.hs:
```diff
@@ -208,7 +209,8 @@ newDBLayer trace fp = do
             let poolMetadataUrl = fst <$> poolMetadata
             let poolMetadataHash = snd <$> poolMetadata
-            insert_ $ PoolRegistration
+            deleteWhere [PoolOwnerPoolId ==. poolId]
```
The delete is redundant with the use of `repsert`, isn't ?

Anviking replied:
No, this is removing the pool _owners_ (before we later `insertMany_ `)

The `repsert` is used with the pool _registration certificate_

Anviking replied:
However — I think my solution here is flawed in the context of rollbacks :/

We need to preserve the old certificate, in case the new certificate gets rolled back.

Maybe we can get away with something simple though. Like just keeping both and reading the latest one, when we need to read it.

KtorZ replied:
Indeed.. rollbacks. What about simply changing the PrimaryKey to:

```
    Primary poolRegistrationPoolId poolRegistrationSlot
```

In the rare case we have two certificates at the same slot, then only the latest one will prevail (we still need the repsert here) as it should, and it'll preserve old certificates on past slots if there are any. 

We'll have to be careful when _reading_ the certificates from the db now, and make sure they're ordered by descending slot. As for the owners :/ ... it's more complicated.

Anviking replied:
> As for the owners :/ ... it's more complicated

Could we not store the owners as some kind of comma-separated list as part of the `pool_registration` table?

Anviking replied:
…or perhaps that's not possible.

KtorZ replied:
:/ ... What about using MongoDB then? 

Not really, this is not quite the problem, and it won't solve the need for a migration. Ideally, we should also attach a slot to the pool owner, or a certificate number, so that when we fetch a certificate, we can fetch _just_ the owner corresponding to that certificate.

Anviking replied:
Are you sure it is possible to use both slot and poolId as primary key, while using them as a foreign key at the same time?

I added (to the `PoolRegistration` table):
```
    Foreign PoolRegistration fk_registration_pool_id poolOwnerPoolId
    Foreign PoolRegistration fk_registration_slot poolOwnerPoolSlot
```

But get:
```

found 1 fkeys and 2 pkeys: fdef=ForeignDef {foreignRefTableHaskell = HaskellName {unHaskellName = "PoolRegistration"}, foreignRefTableDBName = DBName {unDBName = "pool_registration"}, foreignConstraintNameHaskell = HaskellName {unHaskellName = "fk_registration_pool_id"}, foreignConstraintNameDBName = DBName {unDBName = "pool_ownerfk_registration_pool_id"}, foreignFields = [], foreignAttrs = [], foreignNullable = False} pdef=CompositeDef {compositeFields = [FieldDef {fieldHaskell = HaskellName {unHaskellName = "poolRegistrationPoolId"}, fieldDB = DBName {unDBName = "pool_id"}, fieldType = FTTypeCon (Just "W") "PoolId", fieldSqlType = SqlOther "SqlType unset for poolRegistrationPoolId", fieldAttrs = ["sql=pool_id"], fieldStrict = True, fieldReference = NoReference, fieldComments = Nothing},FieldDef {fieldHaskell = HaskellName {unHaskellName = "poolRegistrationSlot"}, fieldDB = DBName {unDBName = "slot"}, fieldType = FTTypeCon (Just "W") "SlotId", fieldSqlType = SqlOther "SqlType unset for poolRegistrationSlot", fieldAttrs = ["sql=slot"], fieldStrict = True, fieldReference = NoReference, fieldComments = Nothing}], compositeAttrs = []}
```

KtorZ replied:
Yes, you need the foreign key to match the primary key of the foreign table. See: https://github.com/input-output-hk/cardano-wallet/pull/1798/commits/81fe0c38b2db9a928c91af52b47729c70cc72438

KtorZ replied:
I think this should do.

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -1866,6 +1875,26 @@ instance Buildable PoolRegistrationCertificate where
+    -- | TODO: Check what this EpochNo actually means.
+    --
+    -- But a guess would be that it is the /first/ epoch that the pool will be
+    -- retired in.
```
Indeed. it should be the epoch at which the pool becomes inactive.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -251,8 +253,17 @@ combineLsqData NodePoolLsqData{nOpt, rewards, stake} =
+    -- TODO: This case seems possible on shelley_testnet, but why, and how
+    -- should we treat it?
+    --
+    -- The pool with rewards but not stake didn't seem to be retiring.
```
Could it be because the leader schedule is based on the stake from the previous epoch? So a pool can have 0 stake now, but had some in the previous epoch and therefore, is getting rewards in this epoch.

Anviking replied:
> So a pool can have 0 stake now

I think that I confirmed that pools in the stake distribution may have 0 stake. I.e. if this pool is entirely _missing_, something else is going on.

KtorZ reviewed (approved):
LGTM

Anviking committed:
fixup: tweak retirement certificate docs

KtorZ committed:
use slot + pool id as primary key for pool registration

This should prevent re-registration to blow up when being stored in the database. There can't possibly be
a pool with two 'valid' certificates on the same slot (the last one would prevail). I had to remove the f
foreign key on the metadata table because it was kinda awkward to link metadata to the pool id in the
first place, but even more awkward to also link the slot.

KtorZ committed:
use TLS-capable HTTP client for fetching metadata from remote peers

Commit:
Regenerate nix
